src/OpenIddict.Core/OpenIddictProvider.Authentication.cs

```diff
@@ -162,13 +177,20 @@ public partial class OpenIddictProvider<TUser, TApplication> : OpenIdConnectServ
             // continue to the next middleware in the pipeline.
             var user = await services.Users.GetUserAsync(principal);
             if (user == null) {
+                logger.LogWarning("Unable to retrieve user from the current principal.");
                 return;
             }
 
             // Note: filtering the username is not needed at this stage as OpenIddictController.Accept
             // and OpenIddictProvider.GrantResourceOwnerCredentials are expected to reject requests that
             // don't include the "email" scope if the username corresponds to the registed email address.
             var identity = await services.Applications.CreateIdentityAsync(user, context.Request.GetScopes());
+
+            if (identity == null) {
+                logger.LogDebug("There was an error during identity creation for user {User}, the current identity is null.", user);
+                // TODO should we fail here witha gracefoul exception other than ArgumentNullException?
```

Comment:
We can (or even should, as one can override `CreateIdentityAsync` to potentially return `null`). I think we should just throw an `InvalidOperationException` at this stage.